File: 2220-find-all-possible-recipes-from-given-supplies/2220-find-all-possible-recipes-from-given-supplies.java
class Solution {
    public List<String> findAllRecipes(String[] recipes, List<List<String>> ingredients, String[] supplies) {
        int n = recipes.length;
        Map<String, ArrayList<String>> edges = new HashMap<>();
        Map<String, Integer> indegree = new HashMap<>();

        for(int i = 0; i<n; i++) {
            indegree.put(recipes[i], ingredients.get(i).size());
            for(String ingredient : ingredients.get(i)) {
                edges.computeIfAbsent(ingredient, k -> new ArrayList<>()).add(recipes[i]);
            }
        }

        Queue<String> pq = new LinkedList<>();
        for(String supply : supplies) {
            pq.add(supply);
        }

        List<String> ans = new ArrayList<>();
        while(!pq.isEmpty()) {
            String item = pq.poll();

            if(indegree.containsKey(item)) {
                ans.add(item);
            }

            for(String recipe : edges.getOrDefault(item, new ArrayList<>())) {
                indegree.put(recipe, indegree.get(recipe) -1);
                if(indegree.get(recipe) == 0) {
                    pq.add(recipe);
                }
            }
        }

        return ans;
    }
}

File: 2220-find-all-possible-recipes-from-given-supplies/2220-find-all-possible-recipes-from-given-supplies.cpp
class Solution {
public:
    vector<string> findAllRecipes(vector<string>& recipes, vector<vector<string>>& ingredients, vector<string>& supplies) {
        int n = recipes.size();
        unordered_map<string, vector<string>> edges;
        unordered_map<string, int> indegree;

        for(int i = 0; i<n; i++) {
            indegree[recipes[i]] = ingredients[i].size();
            for(string ingredient : ingredients[i]) {
                edges[ingredient].push_back(recipes[i]);
            }
        }

        queue<string> pq;
        for(string supply : supplies) {
            pq.push(supply);
        }

        vector<string> ans;
        while(!pq.empty()) {
            string item = pq.front();
            pq.pop();

            if(indegree.find(item) != indegree.end()) {
                ans.push_back(item);
            }

            for(string recipe : edges[item]) {
                indegree[recipe]--;
                if(indegree[recipe] == 0) pq.push(recipe);
            }
        }

        return ans;
    }
};